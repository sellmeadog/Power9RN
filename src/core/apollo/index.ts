export * from './apollo-provider.component';
