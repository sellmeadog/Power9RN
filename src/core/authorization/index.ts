export * from './auth-client';
export * from './authorization.query';
export * from './authorization.service';
