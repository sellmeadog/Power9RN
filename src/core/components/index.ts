export * from './new-app-screen/new-app-screen.component';
