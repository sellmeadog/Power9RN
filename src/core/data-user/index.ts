import { P9DecklistEntryType, P9UserDecklist, P9UserDecklistSchema } from './schema/user-decklist';
import { P9UserDecklistEntrySchema } from './schema/user-decklist-entry';

export { P9DecklistEntryType, P9UserDecklist, P9UserDecklistEntrySchema, P9UserDecklistSchema };
export * from './model/document-info.interface';
