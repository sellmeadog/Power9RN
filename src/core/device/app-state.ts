import { AppState, AppStateStatus } from 'react-native';
import { BehaviorSubject, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { singleton } from 'tsyringe';

@singleton()
export class P9AppState {
  private appState$ = new BehaviorSubject(AppState.currentState);

  readonly active$ = this.appState$.pipe(map((status) => status === 'active'));

  get statusChanges(): Observable<AppStateStatus> {
    return this.appState$.asObservable();
  }

  constructor() {
    AppState.addEventListener('change', (status) => this.appState$.next(status));
  }
}
