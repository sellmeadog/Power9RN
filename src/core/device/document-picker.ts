import { useState } from 'react';
import { Alert } from 'react-native';
import DocumentPicker, { DocumentPickerResponse } from 'react-native-document-picker';
import RNFS from 'react-native-fs';

export type P9DocumentContent = Promise<string>;

export interface P9DocumentInfo extends Pick<DocumentPickerResponse, 'name' | 'type' | 'uri'> {
  content: P9DocumentContent;
}
/**
 * A hook that returns the content of a document picked from the device.
 *
 * @returns An an object that contains the document content as a string and a function to initiate the document picker
 * workflow. Note that `documentInfo` will only be set if a `callback` is NOT provided to the hook.
 */
export const useDocumentPicker = (
  callback?: (documentInfo: P9DocumentInfo) => void,
): [documentInfo: P9DocumentInfo | undefined, pickDocumentAsync: () => Promise<boolean>] => {
  const [documentInfo, setDocumentInfo] = useState<P9DocumentInfo>();

  async function pickDocumentAsync() {
    try {
      const { name, type, uri } = await DocumentPicker.pick({ type: ['public.plain-text'] });
      const [_, inbox] = uri.split('/').reverse();

      const documentPickerResponse: P9DocumentInfo = {
        name,
        type,
        uri,
        content: RNFS.readFile([RNFS.TemporaryDirectoryPath, inbox, name].join('/')).then((content) =>
          content.replace(/\r?\n/g, '\n'),
        ),
      };

      setDocumentInfo(documentPickerResponse);
      callback?.(documentPickerResponse);
      return true;
    } catch (error) {
      if (!DocumentPicker.isCancel(error)) {
        Alert.alert('Document Error', 'An error occurred in the document picker. Please try again.');
      }
      return false;
    }
  }

  return [documentInfo, pickDocumentAsync];
};
