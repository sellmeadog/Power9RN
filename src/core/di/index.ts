export * from './dependency-provider.component';
