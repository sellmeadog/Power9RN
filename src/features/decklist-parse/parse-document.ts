import { P9DocumentInfo } from '../../core/data-user';
import { parseDocumentCsv } from './parse-document-csv';
import { P9CreateDecklistInfo } from './parse-document-types';
import { parseTextEntries } from './parse-text-entries';

export type P9CreateDecklistParseDocumentResult = Pick<
  P9CreateDecklistInfo,
  'name' | 'manualEntries' | 'parsedEntries'
>;

/**
 * Parses the given document and returns a P9CreateDecklistParseDocumentResult
 * @param param0 The P9DocumentInfo object to parse
 */
export const parseDocument = async ({
  name: documentName,
  content,
}: P9DocumentInfo): Promise<P9CreateDecklistParseDocumentResult> => {
  const name = documentName.slice(0, documentName.lastIndexOf('.')).trim();
  const documentContent = await content;

  if (documentName.endsWith('.csv')) {
    return { name, ...parseDocumentCsv(documentContent) };
  } else if (documentName.endsWith('.dec')) {
    // TODO: Parse DEC
    return { name };
  } else {
    return {
      name,
      manualEntries: documentContent,
      parsedEntries: parseTextEntries(documentContent),
    };
  }
};
