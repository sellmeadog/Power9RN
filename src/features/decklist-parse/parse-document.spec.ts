import { P9DocumentInfo } from '../../core/device';
import { parseDocument } from './parse-document';

describe('parseDocument', () => {
  it('should parse a TXT document', async () => {
    // Arrange
    const documentInfo: P9DocumentInfo = {
      name: 'Another One Bites The Dust - Davis.txt',
      content: Promise.resolve(
        `4 Gravecrawler
4 geralfs messenger

3 damping sphere
`,
      ),
      type: 'text/plain',
      uri: 'file://test.txt',
    };

    // Act
    const actual = await parseDocument(documentInfo);

    // Assert
    expect(actual).toMatchSnapshot();
  });

  it('should parse a CSV document', async () => {
    // Arrange
    const documentInfo: P9DocumentInfo = {
      name: 'We Will Rock You - Davis.csv',
      content: Promise.resolve(
        `Count,Tradelist Count,Name,Edition,Card Number,Condition,Language,Foil,Signed,Artist Proof,Altered Art,Mis,
4,4,Angel of Serenity,Return to Ravnica,1,Near Mint,English,,,,,,,,
1,1,Ashen Rider,Theros,187,Near Mint,English,,,,,,,,
1,0,Anax and Cymede,Theros,186,Near Mint,English,foil,,,,,,,
`,
      ),
      type: 'text/csv',
      uri: 'file://test.csv',
    };

    // Act
    const actual = await parseDocument(documentInfo);

    // Assert
    expect(actual).toMatchSnapshot();
  });
});
