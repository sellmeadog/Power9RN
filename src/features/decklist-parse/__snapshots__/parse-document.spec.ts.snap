// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseDocument should parse a CSV document 1`] = `
Object {
  "manualEntries": "4 Angel of Serenity (Return to Ravnica) 1
1 Ashen Rider (Theros) 187
1 Anax and Cymede (Theros) 186",
  "name": "We Will Rock You - Davis",
  "parsedEntries": Array [
    Object {
      "cardName": "Angel of Serenity",
      "collectorNumber": "1",
      "count": "4",
      "expansionCode": "Return to Ravnica",
      "type": "maindeck",
    },
    Object {
      "cardName": "Ashen Rider",
      "collectorNumber": "187",
      "count": "1",
      "expansionCode": "Theros",
      "type": "maindeck",
    },
    Object {
      "cardName": "Anax and Cymede",
      "collectorNumber": "186",
      "count": "1",
      "expansionCode": "Theros",
      "type": "maindeck",
    },
  ],
}
`;

exports[`parseDocument should parse a TXT document 1`] = `
Object {
  "manualEntries": "4 Gravecrawler
4 geralfs messenger

3 damping sphere
",
  "name": "Another One Bites The Dust - Davis",
  "parsedEntries": Array [
    Object {
      "cardName": "Gravecrawler",
      "collectorNumber": undefined,
      "count": "4",
      "expansionCode": undefined,
      "type": "maindeck",
    },
    Object {
      "cardName": "geralfs messenger",
      "collectorNumber": undefined,
      "count": "4",
      "expansionCode": undefined,
      "type": "maindeck",
    },
    Object {
      "cardName": "damping sphere",
      "collectorNumber": undefined,
      "count": "3",
      "expansionCode": undefined,
      "type": "sideboard",
    },
  ],
}
`;
